### PR TITLE
feat: cosmo ai, generate docs on publish

### PR DESCRIPTION
**Pull request by StarpTech**
## Motivation and Context

image

This PR introduces "Cosmo AI" the first step to bring AI capabilities into Cosmo. The user has to opt-in to use it. The first feature is the ability to auto generate subgraph documentation based on the provided schema on `wgc subgraph publish`. 

In order to not block the publish operation for several seconds, we integrated [BullMQ](https://docs.bullmq.io/) a message queue for Node.js and offload the task on a worker. The readme is updated asynchronously after publish.
This also sets the foundation for alerting and recurrent jobs.

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 1]
```diff
@@ -1519,6 +1532,27 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
           targetNamespaceDisplayName: subgraph.namespace,
         });
 
+        if (
+          opts.openaiApiKey &&
+          // Avoid calling OpenAI API if the schema is too big.
+          // Best effort approach. This way of counting tokens is not accurate.
+          subgraph.schemaSDL.length <= 10_000
```
Comment (jensneuse) [inline controlplane/src/core/bufservices/PlatformService.ts:1539]: You can count tokens e.g. using a library like this:
https://github.com/dqbd/tiktoken

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 2]
```diff
@@ -1519,6 +1532,27 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
+        ) {
+          const orgRepo = new OrganizationRepository(opts.db, opts.billingDefaultPlanId);
+          const feature = await orgRepo.getFeature({
```
Comment (jensneuse) [inline controlplane/src/core/bufservices/PlatformService.ts:1542]: what happens if one of these awaits fails?

File: controlplane/src/core/openai-graphql/index.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        frequency_penalty: 0,
+        presence_penalty: 0,
+        stream: false,
+      },
+      {
+        timeout: 60_000,
+      },
+    );
+
+    return { readme: response.choices[0].message.content ?? '' };
```
Comment (jensneuse) [inline controlplane/src/core/openai-graphql/index.ts:47]: What if choices is not an array?

File: controlplane/src/core/workers/AIGraphReadmeWorker.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    },
+  ) {
+    this.oaig = new OpenAIGraphql({
+      openAiApiKey: input.openAiApiKey,
+    });
+    this.input.log = input.log.child({ worker: this.constructor.name });
+  }
+
+  public async handler(job: Job<CreateReadmeInputEvent>) {
+    if (job.data.type === 'subgraph') {
```
Comment (jensneuse) [inline controlplane/src/core/workers/AIGraphReadmeWorker.ts:58]: use switch + move each job to another func?

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: controlplane/src/core/bufservices/PlatformService.ts (hunk 1 above)
Comment (StarpTech) [reply controlplane/src/core/bufservices/PlatformService.ts:1539]: Thanks, I saw it but I'm a bit reluctant to run WASM in our hot paths. It requires more testing.

File: controlplane/src/core/bufservices/PlatformService.ts (hunk 2 above)
Comment (StarpTech) [reply controlplane/src/core/bufservices/PlatformService.ts:1542]: What do you expect? An error is returned to the client.

File: controlplane/src/core/workers/AIGraphReadmeWorker.ts (hunk 4 above)
Comment (StarpTech) [reply controlplane/src/core/workers/AIGraphReadmeWorker.ts:58]: I've moved both into separate functions. I don't see a value here to use switch though.

**Review by jensneuse — APPROVED** [review]

**Pull request merged**